active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+```
+
+An exclusion to this pattern is the following form:
+
+```hbs
+<div {{full-name "Webbigail" "Vanderquack"}}></div>
+```
+
+This is a legacy form of mustache usage. In 2.0 both forms of helpers should
+throw an exception when used in this manner.
+
+### Block form
+
+Helpers can optionally take a template and inverse block. To render a passed
+block, they must yield it. For example, an `if-is-yes` helper might be
+written like so:
+
+```js
+// app/helpers/custom-if.js
+import Ember from "ember";
+
+// Usage: {{if-is-yes 'no'}}not rendered{{else}}rendered!{{/if-is-yes}}
+export default function ifIsYes(params, hash, blocks) {
+  if (params[0] === 'yes') {
+    blocks.template.yield();
+  } else {
+    blocks.inverse.yield();
+  }
+};
+```
+
+Porting this usage to a stateful helper only changes the boilerplate:
+
+```js
+// app/helpers/custom-if.js
+import Ember from "ember";
+
+// Usage: {{if-is-yes 'no'}}not rendered{{else}}rendered!{{/if-is-yes}}
+export default Ember.Helper.extend({
+  compute(params, hash, blocks) {
+    if (params[0]) {
+      blocks.template.yield();
+    } else {
+      blocks.inverse.yield();
+    }
+  }
+});
+```
+
+### Consuming services and recompute
+
+Stateful helpers are a valid target for service injection. For example:
+
+```js
+// app/helpers/current-user-name.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  // Same API as components:
+  session: Ember.inject.service(),
+  compute() {
+    return this.get('session.currentUser.name');
+  }
+});
+```
+
+However consuming a property from a service does not bind the data being
+displayed to that property. After `{{current-user-name}}` has been computed
+and rendered, it will never be invalidated.
+
+For this reason, stateful helpers are granted some control over their
+computation lifecycle. A stateful helper will recompute when:
+
+* A value passed via the template changes (`params` or `hash`)
+* The `recompute` method is called
+* Any key in the `recomputeObserves` array changes
+
+For example, this stateful helper checks if the current use has access to a
+resource type:
+
+```js
+// app/helpers/can-access.js
+import Ember from "ember";
+
+// Usage {{if (can-access 'admin') 'Welcome, boss' 'Heck no!'}}
+export default Ember.Helper.extend({
+  session: Ember.inject.service(),
+  recomputeObserves: ['session.currentUser'],
+  compute(params) {
+    const currentUser = this.get('session.currentUser');
+    return currentUser.can(params[0]);
+  }
+});
+```
+
+`recomputeObserves` can be desugared to:
+
+```js
+  recomputeWhen: Ember.observes('session.currentUser', function() {
+    this.recompute();
+  }),
+```
+
+# Drawbacks
+
+Stateful helpers may superficially appear similar to components, but in
+practice they have none of the special behavior of components. The intent of
+this RFC is that they remain conceptually very close to function helpers,
+but despite this intent they are a new concept for the framework.
```

Comment:
I think we can avoid concept explosion by documenting function helpers as a shorthand for `Ember.Helper`.